File: frontend/src/utils/safeRedirect.js
/**
 * P0 Security Hardening — Safe redirect sanitizer (2026-06)
 * =========================================================
 *
 * Bug class: Open redirect via post-login `?next=` / `?return=` param.
 *
 * Attack surface:
 *   /login?next=https://evil.com         → external phishing
 *   /login?next=//evil.com               → scheme-relative bypass
 *   /login?next=javascript:alert(1)      → XSS via navigation
 *   /login?next=data:text/html,<script>  → data-URL exfil
 *   /login?next=%2F%2Fevil.com           → encoded bypass
 *
 * Contract for a "safe" redirect target:
 *   1. Must be a non-empty string.
 *   2. After exhaustive URL-decoding, must start with `/`.
 *   3. After decoding, must NOT start with `//` (scheme-relative).
 *   4. After decoding, must NOT start with `/\` or `/%5C` (backslash
 *      scheme-relative — Chromium, Edge, Safari all coerce `\` to `/`).
 *   5. Must NOT contain the substring `://` anywhere
 *      (defense-in-depth — catches `/https://evil.com` obfuscation).
 *   6. Must NOT begin with a dangerous scheme (`javascript:`, `data:`,
 *      `vbscript:`, `file:`) — applied to the fully-decoded value.
 *   7. Must NOT loop to `/login` or `/signup`.
 *
 * Anything failing these checks returns the safe fallback:
 *   `/app/dashboard` (founder-mandated post-login landing).
 *
 * Pinned by:
 *   backend/tests/test_safe_redirect_open_redirect_guard_2026_06.py
 */

const SAFE_FALLBACK = '/app/dashboard';

/** Exhaustively url-decode (cap at 5 iterations to prevent pathological loops). */
function exhaustiveDecode(s) {
  let cur = s;
  for (let i = 0; i < 5; i++) {
    let next;
    try {
      next = decodeURIComponent(cur);
    } catch {
      // Malformed encoding — treat the raw value as untrusted.
      return cur;
    }
    if (next === cur) return cur;
    cur = next;
  }
  return cur;
}

/**
 * Return a safe same-origin path or the canonical fallback.
 *
 * @param {string|null|undefined} input - candidate path (decoded or not)
 * @param {string} [fallback='/app/dashboard']
 * @returns {string}
 */
export function safeRedirectPath(input, fallback = SAFE_FALLBACK) {
  if (!input || typeof input !== 'string') return fallback;

  // Trim leading/trailing whitespace AND control characters (some browsers
  // ignore leading whitespace when navigating: `\t//evil.com` could escape).
  // eslint-disable-next-line no-control-regex
  const val = input.replace(/^[\s\u0000-\u001f]+|[\s\u0000-\u001f]+$/g, '');
  if (!val) return fallback;

  // Decode to the canonical form an attacker would actually navigate to.
  const decoded = exhaustiveDecode(val).toLowerCase();
  // Keep an original-case decoded copy for the return value so legitimate
  // paths preserve their casing (e.g., `/app/My-Space`).
  const decodedRaw = exhaustiveDecode(val);

  // Must be a relative same-site path.
  if (!decoded.startsWith('/')) return fallback;
  // Reject scheme-relative `//evil.com`.
  if (decoded.startsWith('//')) return fallback;
  // Reject backslash variants — browsers coerce these to scheme-relative.
  if (decoded.startsWith('/\\') || decoded.startsWith('\\\\')) return fallback;

  // Reject any embedded protocol (defense in depth — catches things like
  // `/https://evil.com` or weird `/foo/bar://evil`).
  if (decoded.includes('://')) return fallback;

  // Reject dangerous schemes right after the leading slash.
  if (/^\/(javascript|data|vbscript|file|about|blob):/i.test(decoded)) {
    return fallback;
  }

  // Prevent self-loops back to auth pages.
  if (
    decoded === '/login' ||
    decoded.startsWith('/login?') ||
    decoded.startsWith('/login#') ||
    decoded.startsWith('/login/') ||
    decoded === '/signup' ||
    decoded.startsWith('/signup?') ||
    decoded.startsWith('/signup#') ||
    decoded.startsWith('/signup/')
  ) {
    return fallback;
  }

  return decodedRaw;
}

/** Exported for tests only. */
export const __SAFE_FALLBACK = SAFE_FALLBACK;

File: backend/utils/safe_redirect.py
"""
P0 SECURITY — Backend redirect sanitizer (2026-06)
====================================================

Server-side equivalent of `frontend/src/utils/safeRedirect.js`. Used by
backend handlers that might ever echo a user-supplied or untrusted
string into a `RedirectResponse(url=...)`, a Cashfree `return_url`, or
any other 3xx Location header.

Current backend audit (2026-06) finds NO active sinks that take
user-controlled redirect input — every existing redirect target is a
server-built string with a hardcoded base path. This module exists
to:

  1. Defensively wrap the `create_subscription(..., return_url=...)`
     parameter so a future regression that passes user input cannot
     turn into an open redirect via Cashfree.
  2. Provide a canonical sanitizer for any future handler that needs
     to accept a `?next=` style param.
  3. Be the bug-class boundary pinned by
     `tests/test_backend_redirect_sink_audit_2026_06.py`.

Contract (mirrors the frontend doctrine)
----------------------------------------
A "safe" redirect target MUST:
  • be a non-empty string
  • after exhaustive URL-decoding, start with `/`
  • NOT start with `//` (scheme-relative)
  • NOT start with `/\\` or `\\\\` (backslash variants)
  • NOT contain `://` anywhere
  • NOT begin with a dangerous scheme after the leading `/`
    (`javascript:`, `data:`, `vbscript:`, `file:`, `about:`, `blob:`)
  • NOT loop to `/login` or `/signup`

Anything failing → canonical fallback `/app/dashboard`.

For Cashfree return URLs (which MUST be absolute https URLs of our
own frontend), there is a separate helper `assert_same_origin_https`
that validates the URL host against an allowlist drawn from the
`FRONTEND_URL`/`ALLOWED_REDIRECT_HOSTS` env vars.
"""
from __future__ import annotations

import os
import re
from urllib.parse import unquote, urlparse


SAFE_FALLBACK = "/app/dashboard"
_DANGEROUS_SCHEMES = re.compile(
    r"^/(javascript|data|vbscript|file|about|blob):",
    re.IGNORECASE,
)
_CONTROL_OR_WS = re.compile(r"^[\s\x00-\x1f]+|[\s\x00-\x1f]+$")


def _exhaustive_decode(s: str, max_passes: int = 5) -> str:
    cur = s
    for _ in range(max_passes):
        try:
            nxt = unquote(cur)
        except Exception:
            return cur
        if nxt == cur:
            return cur
        cur = nxt
    return cur


def safe_redirect_path(value, fallback: str = SAFE_FALLBACK) -> str:
    """Return a safe same-origin path or the canonical fallback.

    Args:
        value: candidate path (decoded or not). Must be str-like.
        fallback: returned for any tampered/unsafe value.
    """
    if not value or not isinstance(value, str):
        return fallback

    stripped = _CONTROL_OR_WS.sub("", value)
    if not stripped:
        return fallback

    decoded_raw = _exhaustive_decode(stripped)
    decoded = decoded_raw.lower()

    # Must be a relative same-site path.
    if not decoded.startswith("/"):
        return fallback
    if decoded.startswith("//"):
        return fallback
    if decoded.startswith("/\\") or decoded.startswith("\\\\"):
        return fallback

    # Any embedded `://` is rejected (defense in depth).
    if "://" in decoded:
        return fallback

    # Dangerous schemes immediately after the leading slash.
    if _DANGEROUS_SCHEMES.match(decoded):
        return fallback

    # Prevent loops back to auth pages.
    for loop_path in ("/login", "/signup"):
        if (
            decoded == loop_path
            or decoded.startswith(loop_path + "?")
            or decoded.startswith(loop_path + "#")
            or decoded.startswith(loop_path + "/")
        ):
            return fallback

    return decoded_raw


# ─────────────────────────────────────────────────────────────────────
# Same-origin HTTPS validation for absolute URLs we hand off to Cashfree.
# ─────────────────────────────────────────────────────────────────────
def _allowed_hosts() -> set:
    """Hosts we will allow as the host of an outbound redirect URL.

    Drawn from env vars so production/staging/preview each lock down to
    their own canonical hosts without code changes.
    """
    hosts = set()
    raw = os.environ.get("ALLOWED_REDIRECT_HOSTS", "")
    for h in raw.split(","):
        h = h.strip().lower()
        if h:
            hosts.add(h)
    # Always allow whatever FRONTEND_URL points at.
    fu = os.environ.get("FRONTEND_URL", "")
    if fu:
        try:
            p = urlparse(fu)
            if p.hostname:
                hosts.add(p.hostname.lower())
        except Exception:
            pass
    # Sensible defaults for the visionary-suite.com prod surface.
    hosts.update({"www.visionary-suite.com", "visionary-suite.com"})
    return hosts


def assert_same_origin_https(url: str) -> str:
    """Validate that `url` is an https URL on an allowed host.

    Used for Cashfree `return_url` / `notify_url` payloads. Raises
    ValueError on violation so the caller fails closed.
    """
    if not isinstance(url, str) or not url:
        raise ValueError("redirect URL is empty")

    decoded = _exhaustive_decode(url.strip())
    parsed = urlparse(decoded)

    if parsed.scheme != "https":
        raise ValueError(
            f"redirect URL must be https, got scheme={parsed.scheme!r}"
        )
    host = (parsed.hostname or "").lower()
    if not host:
        raise ValueError("redirect URL has no host")
    if host not in _allowed_hosts():
        raise ValueError(
            f"redirect URL host {host!r} not in ALLOWED_REDIRECT_HOSTS"
        )
    if parsed.path and (parsed.path.startswith("//") or "://" in parsed.path):
        raise ValueError("redirect URL path looks tampered")
    return decoded


__all__ = [
    "safe_redirect_path",
    "assert_same_origin_https",
    "SAFE_FALLBACK",
]
